### PR TITLE
[WIP] Slightly less ambiitious inference

### PR DESCRIPTION
**Pull request by dlwh**

### CODE REVIEW

**Comment by dlwh** [comment]
@codex can you make a substantive review

File: src/levanter/inference/page_table.py [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+
+    # ------------------------------------------------------------------
+    # Sequence management
+    # ------------------------------------------------------------------
+    @eqx.filter_jit(donate="all")
+    def assign_seq_id_to_seq(self) -> tuple["PageTable", int]:
+        seq_id = hax.argmax(self.seq_lens, "seq").scalar()
+        # Error handling: if there are no sequences available, return INVALID
+        seq_id = hax.where(is_invalid(self.seq_lens["seq", seq_id]), seq_id, INVALID)
+        new_seq_lens = hax.where(
```
[the remark this answers is not in the document]
Comment (dlwh) [reply src/levanter/inference/page_table.py:84]: INVALID is 2000000 you silly bot

**Comment by dlwh** [comment]
@codex can you look again. please review very carefully. there's lots to get right/wrong.

File: src/levanter/models/flash_attention.py [hunk 2]
```diff
@@ -175,6 +175,11 @@ def _flash_attention_forward(
     ell = hax.auto_sharded(ell)
 
     is_causal = isinstance(mask, AttentionMask) and mask.is_causal
+    # Non-zero offsets (dynamic or otherwise) are not supported in this implementation
+    if isinstance(mask, AttentionMask) and mask.is_causal and mask.causal_offset is not None:
+        raise NotImplementedError("Non-zero causal offsets are not implemented for flash attention")
```
Comment (rjpower) [inline src/levanter/models/flash_attention.py:180]: Dumb question, would it make sense for the attention mask to provide `materialize` and `materialize_with_offset`, and then the attention impl would call the appropriate one?

File: src/levanter/data/packing.py [hunk 3]
```diff
@@ -176,6 +176,9 @@ def per_segment_loss(
     assert packed_example.attn_mask.segment_ids is not None, "segment_ids must be set in the AttentionMask"
 
     segment_ids = packed_example.attn_mask.segment_ids
+    if isinstance(segment_ids, tuple):
```
Comment (rjpower) [inline src/levanter/data/packing.py:179]: Is segment_ids always a 1 element tuple in this case?

Comment (rjpower) [reply src/levanter/data/packing.py:179]: (should we always keep it as a tuple to avoid the checks?)

File: src/levanter/layers/attention.py [hunk 4]
```diff
@@ -1325,38 +1474,37 @@ def init(config: AttentionConfig, *, key) -> "Attention":
 
         return Attention(config, q_proj, k_proj, v_proj, o_proj, q_norm, k_norm, rot_embs)
 
+    def empty_cache(self, Batch: Axis, MaxLen: Axis, *, dtype):
```
Comment (rjpower) [inline src/levanter/layers/attention.py]: nit empty_kv_cache for consistency?

File: src/levanter/layers/attention.py [hunk 5]
```diff
@@ -1374,18 +1522,654 @@ def __call__(
...
+
+    is_special_case = jnp.all(new_lengths <= 1)
+
+    updated_keys, updated_values = jax.lax.cond(is_special_case, specialized_path, general_path, operands)
+    updated_lengths = lengths + new_lengths
+
+    return updated_keys, updated_values, updated_lengths
+
+
+class KvPageCache(eqx.Module):
```
Comment (rjpower) [inline src/levanter/layers/attention.py:1626]: I have I had some trouble understanding the logic for the kvcache vs kvpagecache. I naively expected them to share some interface, but it seems like they're completely separate beasts instead.

Not actionable except as a general thought for how maybe paged decoding could be clearly separated out.

File: src/levanter/layers/attention.py [hunk 6]
```diff
@@ -784,39 +839,79 @@ def materialize(
...
             sliding_window: If provided, restrict each query position to attend only to keys within
                 ``sliding_window`` previous positions.
+            For ``offset == 0`` this is identical to the old ``AttentionMask.causal()``
+            behaviour; larger offsets loosen the restriction so that each query can
+            see ``offset`` additional future tokens.
         """
-        return AttentionMask(is_causal=True, sliding_window=sliding_window)
+        if isinstance(offset, int | Integral):
+            causal_offset = hax.named(offset, ())
+        else:
```
[the remark this answers is not in the document]
Comment (dlwh) [reply src/levanter/layers/attention.py:857]: fake news

File: src/levanter/models/flash_attention.py (hunk 2 above)
Comment (dlwh) [reply src/levanter/models/flash_attention.py:180]: hrm, well there's kind of two different notions of offset one might want here. materialize optionally takes a slice of both qpos and kpos and will slice the matrix that way. but you can have other offsets too. I dunno. maybe your idea is the right thing.

File: src/levanter/data/packing.py (hunk 3 above)
Comment (dlwh) [reply src/levanter/data/packing.py:179]: yeah probably. was trying to preserve some compat but it's mostly a pain

Comment (dlwh) [reply src/levanter/data/packing.py:179]: will do a follow up pr

File: src/levanter/layers/attention.py (hunk 4 above)
Comment (dlwh) [reply src/levanter/layers/attention.py]: i'm gonna delete this one actually

File: src/levanter/layers/attention.py (hunk 5 above)
Comment (dlwh) [reply src/levanter/layers/attention.py:1626]: i'm gonna delete pagecache

Comment (dlwh) [reply src/levanter/layers/attention.py:1626]: err, i mean kvcache
